<a href="https://colab.research.google.com/github/fatin-sheyab1/call-transcription-api/blob/main/PIIExtraction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [8]:
!pip install -q datasets

from datasets import load_dataset

data = load_dataset(
    "ai4privacy/pii-masking-200k",
    data_files="english_pii_43k.jsonl"
)

README.md:   0%|          | 0.00/14.1k [00:00<?, ?B/s]

english_pii_43k.jsonl: reconstructing file:   0%|          |  0.00B / 73.8MB            

english_pii_43k.jsonl: downloading bytes:           |  0.00B            

Generating train split: 0 examples [00:00, ? examples/s]

In [9]:
data


DatasetDict({
    train: Dataset({
        features: ['source_text', 'target_text', 'privacy_mask', 'span_labels', 'mbert_text_tokens', 'mbert_bio_labels', 'id', 'language', 'set'],
        num_rows: 43501
    })
})

In [10]:
dataset = data["train"]

print(dataset)

Dataset({
    features: ['source_text', 'target_text', 'privacy_mask', 'span_labels', 'mbert_text_tokens', 'mbert_bio_labels', 'id', 'language', 'set'],
    num_rows: 43501
})


In [11]:
print(dataset)
print("\nColumns:")
print(dataset.column_names)
print("\nNumber of rows : ",len(dataset))

Dataset({
    features: ['source_text', 'target_text', 'privacy_mask', 'span_labels', 'mbert_text_tokens', 'mbert_bio_labels', 'id', 'language', 'set'],
    num_rows: 43501
})

Columns:
['source_text', 'target_text', 'privacy_mask', 'span_labels', 'mbert_text_tokens', 'mbert_bio_labels', 'id', 'language', 'set']

Number of rows :  43501


In [12]:
from collections import Counter
print("Language :")
print(Counter(dataset["language"]))

print("\nSet Values :")
print(Counter(dataset["set"]))

Language :
Counter({'en': 43501})

Set Values :
Counter({'train': 43501})


In [13]:
dataset.select(range(5)).to_pandas()

,source_text,target_text,privacy_mask,span_labels,mbert_text_tokens,mbert_bio_labels,id,language,set
0,A student's assessment was found on device bea...,A student's assessment was found on device bea...,"[{'value': '06-184755-866851-3', 'start': 57, ...","[[0, 57, ""O""], [57, 75, ""PHONEIMEI""], [75, 138...","[A, student, ', s, assessment, was, found, on,...","[O, O, O, O, O, O, O, O, O, O, O, O, O, O, B-P...",165761,en,train
1,"Dear Omer, as per our records, your license 78...","Dear [FIRSTNAME], as per our records, your lic...","[{'value': 'Omer', 'start': 5, 'end': 9, 'labe...","[[0, 5, ""O""], [5, 9, ""FIRSTNAME""], [9, 44, ""O""...","[Dear, Omer, ,, as, per, our, records, ,, your...","[O, B-FIRSTNAME, O, O, O, O, O, O, O, O, B-VEH...",165762,en,train
2,Kattie could you please share your recomndatio...,[FIRSTNAME] could you please share your recomn...,"[{'value': 'Kattie', 'start': 0, 'end': 6, 'la...","[[0, 6, ""FIRSTNAME""], [6, 75, ""O""], [75, 77, ""...","[Kat, ##tie, could, you, pl, ##eas, ##e, share...","[B-FIRSTNAME, I-FIRSTNAME, O, O, O, O, O, O, O...",165763,en,train
3,Emergency supplies in 16356 need a refill. Use...,Emergency supplies in [BUILDINGNUMBER] need a ...,"[{'value': '16356', 'start': 22, 'end': 27, 'l...","[[0, 22, ""O""], [22, 27, ""BUILDINGNUMBER""], [27...","[Emergency, supplies, in, 1635, ##6, need, a, ...","[O, O, O, B-BUILDINGNUMBER, I-BUILDINGNUMBER, ...",165764,en,train
4,"The 88 old child at 5862, has showcased an unu...","The [AGE] old child at [BUILDINGNUMBER], has s...","[{'value': '88', 'start': 4, 'end': 6, 'label'...","[[0, 4, ""O""], [4, 6, ""AGE""], [6, 20, ""O""], [20...","[The, 88, old, child, at, 586, ##2, ,, has, sh...","[O, B-AGE, O, O, O, B-BUILDINGNUMBER, I-BUILDI...",165765,en,train


In [14]:
for column in dataset.column_names:
 missing_count = sum (
     value is None
     for value in dataset[column]

 )

 print(column,":",missing_count)

source_text : 0
target_text : 0
privacy_mask : 0
span_labels : 0
mbert_text_tokens : 0
mbert_bio_labels : 0
id : 0
language : 0
set : 0


In [15]:
source_text = dataset["source_text"]
duplicate_count = len(source_text) - len(set(source_text))

print("Duplicate source_text rows :" ,duplicate_count)

Duplicate source_text rows : 0


In [16]:
from collections import Counter

label_counts = Counter(
    label
    for labels in dataset["mbert_bio_labels"]
    for label in labels
)

print(label_counts)

Counter({'O': 1237294, 'I-USERAGENT': 85005, 'I-IPV6': 71631, 'I-BITCOINADDRESS': 64552, 'I-ETHEREUMADDRESS': 49047, 'I-EMAIL': 36812, 'I-IP': 36649, 'I-URL': 30358, 'I-IBAN': 27821, 'I-PASSWORD': 24089, 'I-DATE': 21392, 'I-CREDITCARDNUMBER': 20716, 'I-LITECOINADDRESS': 20255, 'I-PHONEIMEI': 18858, 'I-NEARBYGPSCOORDINATE': 17660, 'I-MASKEDNUMBER': 17094, 'I-IPV4': 16277, 'I-PHONENUMBER': 14402, 'I-MAC': 14135, 'B-FIRSTNAME': 13293, 'I-COMPANYNAME': 11616, 'I-JOBTITLE': 11527, 'I-USERNAME': 11219, 'I-DOB': 10843, 'I-VEHICLEVIN': 10139, 'I-AMOUNT': 9797, 'I-SSN': 9560, 'I-ACCOUNTNUMBER': 9113, 'I-FIRSTNAME': 9083, 'I-ZIPCODE': 6476, 'I-LASTNAME': 5791, 'I-STREET': 5532, 'I-TIME': 5419, 'I-ACCOUNTNAME': 5315, 'I-CITY': 5264, 'I-GENDER': 5202, 'I-BIC': 5055, 'B-LASTNAME': 4803, 'I-SECONDARYADDRESS': 4591, 'B-DATE': 4532, 'I-VEHICLEVRM': 4114, 'B-EMAIL': 4067, 'I-CURRENCY': 3764, 'B-PREFIX': 3440, 'I-COUNTY': 3042, 'B-USERNAME': 3035, 'B-JOBTITLE': 2930, 'B-URL': 2902, 'B-MIDDLENAME': 2877,

In [17]:
mismatched_dataset = dataset.filter(
    lambda row:
    len(row["mbert_text_tokens"]) != len(row["mbert_bio_labels"])
)

print("Mismatched rows :",len(mismatched_dataset))


Filter:   0%|          | 0/43501 [00:00<?, ? examples/s]

Mismatched rows : 0


In [18]:
all_labels = set(
    label
    for labels in dataset["mbert_bio_labels"]
    for label in labels
)
invalid_labels= [
    label for label in all_labels
    if not (label == "O" or label.startswith("B-") or label.startswith("I-"))
]

print("Invalid labels :",invalid_labels)

Invalid labels : []


In [19]:
def has_invalid_bio_sequence(labels):
    previous_type = None

    for label in labels:

        if label == "O":
            previous_type = None

        elif label.startswith("B-"):
            previous_type = label[2:]

        elif label.startswith("I-"):
            current_type = label[2:]

            if previous_type != current_type:
                return True

            previous_type = current_type

    return False


invalid_bio_dataset = dataset.filter(
    lambda row:
    has_invalid_bio_sequence(row["mbert_bio_labels"])
)

print("Invalid BIO sequence rows :",len(invalid_bio_dataset))


Filter:   0%|          | 0/43501 [00:00<?, ? examples/s]

Invalid BIO sequence rows : 608


In [20]:
print(
    invalid_bio_dataset.select(range(min(10, len(invalid_bio_dataset))))
    .to_pandas()[
        ["mbert_text_tokens", "mbert_bio_labels"]
    ].to_string()
)

                                                                                                                                                                                                                                                                                                                                                                                                              mbert_text_tokens                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 mbert_bio_labels
0                      

In [21]:
row = dataset[68]

print("SOURCE TEXT:")
print(row["source_text"])

print("\nSPAN LABELS:")
print(row["span_labels"])

print("\nMBERT TOKENS:")
print(row["mbert_text_tokens"])

print("\nMBERT BIO LABELS:")
print(row["mbert_bio_labels"])

SOURCE TEXT:
Please allocate digital resources for 12 years old students accessing courses online. Approx budget is Belize Dollar﷼868k. Payment will be made from 10972500.

SPAN LABELS:
[[0, 38, "O"], [38, 50, "AGE"], [50, 103, "O"], [103, 116, "CURRENCY"], [116, 117, "CURRENCYSYMBOL"], [117, 121, "AMOUNT"], [121, 149, "O"], [149, 157, "ACCOUNTNUMBER"], [157, 158, "O"]]

MBERT TOKENS:
['Please', 'allo', '##cate', 'digital', 'resources', 'for', '12', 'years', 'old', 'students', 'access', '##ing', 'courses', 'online', '.', 'App', '##rox', 'budget', 'is', 'Belize', '[UNK]', '.', 'Pay', '##ment', 'will', 'be', 'made', 'from', '1097', '##25', '##00', '.']

MBERT BIO LABELS:
['O', 'O', 'O', 'O', 'O', 'O', 'B-AGE', 'I-AGE', 'I-AGE', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'B-CURRENCY', 'I-CURRENCY', 'I-CURRENCY', 'B-CURRENCYSYMBOL', 'I-AMOUNT', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O']


I found 608 records with invalid BIO label sequences. I tried to fix them, but tokenization and alignment issues could introduce incorrect labels, so I excluded these records from model training to avoid label noise.

In [22]:
clean_dataset = dataset.filter(
    lambda row:
    not has_invalid_bio_sequence(row["mbert_bio_labels"])
)




Filter:   0%|          | 0/43501 [00:00<?, ? examples/s]

In [23]:
print("Original data :",len(dataset))
print("Excluded rows :",len(dataset)-len(clean_dataset))
print("Clean data :",len(clean_dataset))

Original data : 43501
Excluded rows : 608
Clean data : 42893


In [24]:
first_split = clean_dataset.train_test_split(
    test_size=0.2,
    seed=42
)

train_dataset = first_split["train"]
temp_dataset = first_split["test"]

second_split = temp_dataset.train_test_split(
    test_size=0.5,
    seed=42
)

val_dataset = second_split["train"]
test_dataset = second_split["test"]

print("Train:", len(train_dataset))
print("Validation:", len(val_dataset))
print("Test:", len(test_dataset))

Train: 34314
Validation: 4289
Test: 4290


In [25]:
all_labels = sorted(set(
    label
    for labels in train_dataset["mbert_bio_labels"]
    for label in labels
))

print("Number of labels:", len(all_labels))
print(all_labels)

Number of labels: 113
['B-ACCOUNTNAME', 'B-ACCOUNTNUMBER', 'B-AGE', 'B-AMOUNT', 'B-BIC', 'B-BITCOINADDRESS', 'B-BUILDINGNUMBER', 'B-CITY', 'B-COMPANYNAME', 'B-COUNTY', 'B-CREDITCARDCVV', 'B-CREDITCARDISSUER', 'B-CREDITCARDNUMBER', 'B-CURRENCY', 'B-CURRENCYCODE', 'B-CURRENCYNAME', 'B-CURRENCYSYMBOL', 'B-DATE', 'B-DOB', 'B-EMAIL', 'B-ETHEREUMADDRESS', 'B-EYECOLOR', 'B-FIRSTNAME', 'B-GENDER', 'B-HEIGHT', 'B-IBAN', 'B-IP', 'B-IPV4', 'B-IPV6', 'B-JOBAREA', 'B-JOBTITLE', 'B-JOBTYPE', 'B-LASTNAME', 'B-LITECOINADDRESS', 'B-MAC', 'B-MASKEDNUMBER', 'B-MIDDLENAME', 'B-NEARBYGPSCOORDINATE', 'B-ORDINALDIRECTION', 'B-PASSWORD', 'B-PHONEIMEI', 'B-PHONENUMBER', 'B-PIN', 'B-PREFIX', 'B-SECONDARYADDRESS', 'B-SEX', 'B-SSN', 'B-STATE', 'B-STREET', 'B-TIME', 'B-URL', 'B-USERAGENT', 'B-USERNAME', 'B-VEHICLEVIN', 'B-VEHICLEVRM', 'B-ZIPCODE', 'I-ACCOUNTNAME', 'I-ACCOUNTNUMBER', 'I-AGE', 'I-AMOUNT', 'I-BIC', 'I-BITCOINADDRESS', 'I-BUILDINGNUMBER', 'I-CITY', 'I-COMPANYNAME', 'I-COUNTY', 'I-CREDITCARDCVV', 'I-CR

#هون حولت ال labels لارقام بعدين عملت العكس عشان لما يعمل تنبؤ ما يعطيني رقم واكون مش فاهمه عليه يعطي ال label

In [26]:
label2id = {label:i for i,label in enumerate(all_labels)}

id2label = {i: label for label, i in label2id.items()}

print(label2id)
print("\nNumber of labels :",len(label2id))

{'B-ACCOUNTNAME': 0, 'B-ACCOUNTNUMBER': 1, 'B-AGE': 2, 'B-AMOUNT': 3, 'B-BIC': 4, 'B-BITCOINADDRESS': 5, 'B-BUILDINGNUMBER': 6, 'B-CITY': 7, 'B-COMPANYNAME': 8, 'B-COUNTY': 9, 'B-CREDITCARDCVV': 10, 'B-CREDITCARDISSUER': 11, 'B-CREDITCARDNUMBER': 12, 'B-CURRENCY': 13, 'B-CURRENCYCODE': 14, 'B-CURRENCYNAME': 15, 'B-CURRENCYSYMBOL': 16, 'B-DATE': 17, 'B-DOB': 18, 'B-EMAIL': 19, 'B-ETHEREUMADDRESS': 20, 'B-EYECOLOR': 21, 'B-FIRSTNAME': 22, 'B-GENDER': 23, 'B-HEIGHT': 24, 'B-IBAN': 25, 'B-IP': 26, 'B-IPV4': 27, 'B-IPV6': 28, 'B-JOBAREA': 29, 'B-JOBTITLE': 30, 'B-JOBTYPE': 31, 'B-LASTNAME': 32, 'B-LITECOINADDRESS': 33, 'B-MAC': 34, 'B-MASKEDNUMBER': 35, 'B-MIDDLENAME': 36, 'B-NEARBYGPSCOORDINATE': 37, 'B-ORDINALDIRECTION': 38, 'B-PASSWORD': 39, 'B-PHONEIMEI': 40, 'B-PHONENUMBER': 41, 'B-PIN': 42, 'B-PREFIX': 43, 'B-SECONDARYADDRESS': 44, 'B-SEX': 45, 'B-SSN': 46, 'B-STATE': 47, 'B-STREET': 48, 'B-TIME': 49, 'B-URL': 50, 'B-USERAGENT': 51, 'B-USERNAME': 52, 'B-VEHICLEVIN': 53, 'B-VEHICLEVRM'

In [27]:
!pip install -q transformers datasets accelerate seqeval

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.6/43.6 kB 2.3 MB/s eta 0:00:00
  error: subprocess-exited-with-error
  
  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  Preparing metadata (setup.py) ... error
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.


In [28]:
from transformers import AutoTokenizer,AutoModelForTokenClassification

In [29]:
model_name = "google-bert/bert-base-multilingual-cased"

In [30]:
tokenizer = AutoTokenizer.from_pretrained(model_name)

config.json:   0%|          | 0.00/625 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/49.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/996k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.96M [00:00<?, ?B/s]

In [31]:
model = AutoModelForTokenClassification.from_pretrained(
    model_name,
    num_labels = len(all_labels),
    id2label = id2label,
    label2id = label2id

)

model.safetensors: reconstructing file:   0%|          |  0.00B /  714MB            

model.safetensors: downloading bytes:           |  0.00B            

Loading weights:   0%|          | 0/197 [00:00<?, ?it/s]

[transformers] BertForTokenClassification LOAD REPORT from: google-bert/bert-base-multilingual-cased
Key                                        | Status     | 
-------------------------------------------+------------+-
cls.predictions.transform.LayerNorm.bias   | UNEXPECTED | 
cls.seq_relationship.bias                  | UNEXPECTED | 
cls.predictions.transform.dense.weight     | UNEXPECTED | 
bert.pooler.dense.bias                     | UNEXPECTED | 
bert.pooler.dense.weight                   | UNEXPECTED | 
cls.predictions.transform.dense.bias       | UNEXPECTED | 
cls.seq_relationship.weight                | UNEXPECTED | 
cls.predictions.bias                       | UNEXPECTED | 
cls.predictions.transform.LayerNorm.weight | UNEXPECTED | 
classifier.bias                            | MISSING    | 
classifier.weight                          | MISSING    | 

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.
- MISSING:	

In [32]:
max_length = 512

def prepare_example(example):
    tokens = example["mbert_text_tokens"]
    bio_labels = example["mbert_bio_labels"]

    tokens = tokens[:max_length - 2]
    bio_labels = bio_labels[:max_length - 2]

    token_ids = tokenizer.convert_tokens_to_ids(tokens)
    input_ids = (
        [tokenizer.cls_token_id]
        + token_ids
        +[tokenizer.sep_token_id]
    )


    label_ids = [label2id[label] for label in bio_labels]

    labels = [-100] + label_ids + [-100]

    attention_mask = [1] * len(input_ids)

    return {
        "input_ids": input_ids,
        "attention_mask": attention_mask,
        "labels": labels
    }

In [33]:
train_encoded = train_dataset.map(prepare_example)
val_encoded = val_dataset.map(prepare_example)
test_encoded = test_dataset.map(prepare_example)

Map:   0%|          | 0/34314 [00:00<?, ? examples/s]

Map:   0%|          | 0/4289 [00:00<?, ? examples/s]

Map:   0%|          | 0/4290 [00:00<?, ? examples/s]

In [34]:
print(len(train_encoded[0]["input_ids"]))
print(len(train_encoded[0]["labels"]))


27
27


#هسا بدي اعمل اشي اسمه  padding لانه مش كل الصفوف عندي ياهم نفس الطول وهي مشكله اثناء تدريب الموديل لازم يكونو نفس الطول لانه احنا ما مناخد صف صف انما مجموعه من الصفوف اسمها (batch)
-------------------------------------
row 1 :[CLS] Hello [SEP]
row 2 :[CLS] My name is Fatin [SEP]

Row 1 :[CLS] Hello [SEP] [PAD] [PAD] [PAD]
Row 2: [CLS] My name is Ahmad [SEP]

PAD = 0

In [35]:
from transformers import DataCollatorForTokenClassification

data_collator = DataCollatorForTokenClassification(
    tokenizer = tokenizer
)

In [36]:
!pip install -q --upgrade pip setuptools wheel

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 68.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 818.2/818.2 kB 41.1 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
ipython 7.34.0 requires jedi>=0.16, which is not installed.
torch 2.11.0+cu128 requires setuptools<82, but you have setuptools 84.0.0 which is incompatible.


In [37]:
!pip install -q seqeval --no-build-isolation

  Preparing metadata (pyproject.toml) ... done
  Preparing metadata (pyproject.toml) ... done
  Preparing metadata (pyproject.toml) ... done
  Preparing metadata (pyproject.toml) ... done
  Preparing metadata (pyproject.toml) ... done
  Preparing metadata (pyproject.toml) ... done
  Preparing metadata (pyproject.toml) ... done
  Preparing metadata (pyproject.toml) ... done
  Preparing metadata (pyproject.toml) ... done
  Preparing metadata (pyproject.toml) ... done
  Preparing metadata (pyproject.toml) ... done
  Preparing metadata (pyproject.toml) ... done
  Preparing metadata (pyproject.toml) ... done
  Preparing metadata (pyproject.toml) ... done


In [38]:
from seqeval.metrics import precision_score, recall_score, f1_score

In [48]:
import numpy as np
from seqeval.metrics import accuracy_score

def compute_matrix(eval_pred):
  predictions,labels = eval_pred
  predictions = np.argmax(predictions,axis = 2)

  true_predictions =[]
  true_labels=[]
  for prediction ,label in zip(predictions,labels):
    pred_list = []
    label_list = []
    for pred_id,label_id in zip(prediction,label):
      if label_id != -100:
        pred_list.append(id2label[pred_id])
        label_list.append(id2label[label_id])
    true_predictions.append(pred_list)
    true_labels.append(label_list)
  return{
      "accuracy": accuracy_score(true_labels, true_predictions),
      "precision":precision_score(true_labels,true_predictions),
      "recall":recall_score(true_labels,true_predictions),
      "f1":f1_score(true_labels,true_predictions)

  }

In [40]:
from transformers import TrainingArguments,Trainer

training_args = TrainingArguments(
    output_dir="./pii_mbert_model",
    learning_rate=2e-5,
    per_device_train_batch_size=8,
    per_device_eval_batch_size=8,
    num_train_epochs=3,
    weight_decay=0.01,
    eval_strategy="epoch",
    save_strategy="epoch",
    logging_strategy="steps",
    logging_steps=100,
    load_best_model_at_end=True,
    metric_for_best_model="f1",
    greater_is_better=True
)

In [41]:
trainer = Trainer(
    model = model,
    args = training_args,
    train_dataset = train_encoded,
    eval_dataset = val_encoded,
    data_collator = data_collator,
    compute_metrics = compute_matrix,

)

In [42]:
trainer.train()

Epoch,Training Loss,Validation Loss,Precision,Recall,F1
1,0.078493,0.080427,0.921968,0.937547,0.929692
2,0.075639,0.073792,0.936878,0.953476,0.945104
3,0.063473,0.071532,0.936616,0.955858,0.946139


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

[transformers] There were missing keys in the checkpoint model loaded: ['bert.embeddings.LayerNorm.weight', 'bert.embeddings.LayerNorm.bias', 'bert.encoder.layer.0.attention.output.LayerNorm.weight', 'bert.encoder.layer.0.attention.output.LayerNorm.bias', 'bert.encoder.layer.0.output.LayerNorm.weight', 'bert.encoder.layer.0.output.LayerNorm.bias', 'bert.encoder.layer.1.attention.output.LayerNorm.weight', 'bert.encoder.layer.1.attention.output.LayerNorm.bias', 'bert.encoder.layer.1.output.LayerNorm.weight', 'bert.encoder.layer.1.output.LayerNorm.bias', 'bert.encoder.layer.2.attention.output.LayerNorm.weight', 'bert.encoder.layer.2.attention.output.LayerNorm.bias', 'bert.encoder.layer.2.output.LayerNorm.weight', 'bert.encoder.layer.2.output.LayerNorm.bias', 'bert.encoder.layer.3.attention.output.LayerNorm.weight', 'bert.encoder.layer.3.attention.output.LayerNorm.bias', 'bert.encoder.layer.3.output.LayerNorm.weight', 'bert.encoder.layer.3.output.LayerNorm.bias', 'bert.encoder.layer.4.atte

TrainOutput(global_step=12870, training_loss=0.10701046864376942, metrics={'train_runtime': 2084.467, 'train_samples_per_second': 49.385, 'train_steps_per_second': 6.174, 'total_flos': 4456385699840172.0, 'train_loss': 0.10701046864376942, 'epoch': 3.0})

In [43]:
test_result = trainer.evaluate(test_encoded)
print(test_result)

Training Loss,Validation Loss,Epoch,Precision,Recall,F1
0.063473,0.067719,3,0.938025,0.951835,0.944880


{'eval_loss': 0.06771878898143768, 'eval_precision': 0.9380254445353768, 'eval_recall': 0.9518345160803261, 'eval_f1': 0.9448795293588639}


In [44]:
save_path = "/content/drive/MyDrive/pii_mbert_final"

trainer.save_model(save_path)
tokenizer.save_pretrained(save_path)

print("Model Saved Successfully to Google Drive")

Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

Model Saved Successfully to Google Drive


In [45]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [47]:
import os

model_path = "/content/drive/MyDrive/pii_mbert_final"

print(os.listdir(model_path))

['config.json', 'model.safetensors', 'tokenizer_config.json', 'tokenizer.json', 'training_args.bin']


In [49]:
trainer.compute_metrics = compute_matrix

val_result = trainer.evaluate(val_encoded)

test_result = trainer.evaluate(test_encoded)


print("Validation :",val_result)
print("Test :",test_result)

Training Loss,Validation Loss,Epoch,Accuracy,Precision,Recall,F1
0.063473,0.071532,3,0.973627,0.936616,0.955858,0.946139


Training Loss,Validation Loss,Epoch,Accuracy,Precision,Recall,F1
0.063473,0.067719,3,0.974139,0.938025,0.951835,0.944880


Validation : {'eval_loss': 0.07153192162513733, 'eval_accuracy': 0.9736274022847973, 'eval_precision': 0.936615609044493, 'eval_recall': 0.9558582700610392, 'eval_f1': 0.9461391099322134}
Test : {'eval_loss': 0.06771878898143768, 'eval_accuracy': 0.9741389601149804, 'eval_precision': 0.9380254445353768, 'eval_recall': 0.9518345160803261, 'eval_f1': 0.9448795293588639}


In [60]:
!pip install -q fastapi uvicorn nest-asyncio pyngrok

In [67]:
from fastapi import FastAPI
from pydantic import BaseModel
import torch
from transformers import AutoTokenizer, AutoModelForTokenClassification

app = FastAPI()
model_path = "/content/drive/MyDrive/pii_mbert_final"
tokenizer = AutoTokenizer.from_pretrained(model_path)
model = AutoModelForTokenClassification.from_pretrained(model_path)

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model.to(device)
model.eval()

class TranscriptRequest(BaseModel):
    transcript: str


def predict_pii(text):
    inputs = tokenizer(
        text,
        return_tensors="pt",
        truncation=True,
        max_length=512
    )

    inputs = {key: value.to(device) for key, value in inputs.items()}

    with torch.no_grad():
        outputs = model(**inputs)

    predictions = torch.argmax(outputs.logits, dim=2)

    tokens = tokenizer.convert_ids_to_tokens(
        inputs["input_ids"][0]
    )

    labels = [
        model.config.id2label[pred.item()]
        for pred in predictions[0]
    ]

    return tokens, labels


@app.get("/")
def home():
    return {
        "message": "PII Extraction API is running"
    }

Loading weights:   0%|          | 0/199 [00:00<?, ?it/s]

In [70]:
def extract_entities(tokens, labels):
    entities = []
    current_type = None
    current_value = ""

    for token, label in zip(tokens, labels):

        if label == "O":
            if current_value:
                entities.append({
                    "type": current_type,
                    "value": current_value
                })
                current_type = None
                current_value = ""
            continue

        entity_type = label[2:]

        if label.startswith("B-"):
            if current_value:
                entities.append({
                    "type": current_type,
                    "value": current_value
                })

            current_type = entity_type
            current_value = token.replace("##", "")

        elif label.startswith("I-") and current_type == entity_type:
            if token.startswith("##"):
                current_value += token[2:]
            else:
                current_value += token

    if current_value:
        entities.append({
            "type": current_type,
            "value": current_value
        })

    return entities

In [71]:
@app.post("/extract_pii")
def extract_pii(request:TranscriptRequest):
  tokens ,labels = predict_pii(request.transcript)
  entities = extract_entities(tokens,labels)
  return {
      "pii":entities
      }


In [ ]:
import uvicorn

config = uvicorn.Config(
    app,
    host="0.0.0.0",
    port=8000
)

server = uvicorn.Server(config)

await server.serve()

INFO:     Started server process [1950]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)
